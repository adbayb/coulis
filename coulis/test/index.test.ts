import { beforeAll, describe, expect, test } from "vitest";

import {
	createKeyframes,
	createStyles,
	createVariants,
	extractStyles,
	setGlobalStyles,
} from "../src";

describe("coulis", () => {
	beforeAll(() => {
		setGlobalStyles({
			"*,*::before,*::after": {
				boxSizing: "inherit",
			},
			".globalClass+.otherGlobalClass": {
				border: "1px solid black",
				borderRadius: 4,
			},
			"@charset": '"utf-8"',
			"@font-face": {
				fontFamily: "'AliasedHelvetica'",
				src: "local(Helvetica)",
			},
			"@import":
				"url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap')",
			"html": {
				boxSizing: "border-box",
			},
			"html,body": {
				backgroundColor: "lightcoral",
				fontFamily: "Open Sans, AliasedHelvetica",
				margin: 0,
				padding: 0,
			},
		});
	});

	test("should generate classNames", () => {
		expect(animationName).toBe("c77e20e50");
		expect(buttonVariants({ color: "brand", size: "large" })).toBe(
			"c72f7daf8 c6a0088d1",
		);
		expect(buttonVariants({ color: "brand", size: "medium" })).toBe(
			"c72f7daf8 ccadc8a20",
		);
		expect(buttonVariants({ color: "neutral", size: "medium" })).toBe(
			"c5ea183e8 ccadc8a20",
		);
		expect(buttonVariants({ color: "accent", size: "small" })).toBe(
			"cbb3938 c2a100aa4",
		);
		expect(classNames).toBe(
			"cdd299de8 cbb3938 ce31a3799 c18a19c73 c2a100aa4",
		);
	});

	test("should extract styles given object format", () => {
		expect(extractStyles()).toMatchSnapshot();
	});

	test("should extract styles given string format", () => {
		// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
		expect(`${extractStyles()}`).toMatchSnapshot();
	});

	test("should type `createStyles` in a safe manner", () => {
		const styles = createStyles(
			{
				backgroundColor: ["red", "blue"],
				color: {
					danger: "red",
					warning: "yellow",
				},
				height: ["100%"],
				transitionProperty(input: ("background-color" | "color")[]) {
					return input.join(" ");
				},
				width: true,
			},
			{
				loose: ["color"],
				shorthands: {
					size: ["width"],
				},
				states: {
					hover: ({ className, declaration }) =>
						`${className}:hover{${declaration}}`,
				},
			},
		);

		expect(
			styles({
				// @ts-expect-error property value does not exist (custom value)
				backgroundColor: "blue2",
				transitionProperty: ["color", "background-color"],
			}),
		).toBeTypeOf("string");

		expect(
			styles({
				backgroundColor: {
					// @ts-expect-error stateful property value does not exist (custom value)
					base: "blue3",
				},
			}),
		).toBeTypeOf("string");

		expect(() =>
			styles({
				backgroundColor: {
					// @ts-expect-error state key does not exist
					toto: "red",
				},
			}),
		).toThrow(/No configuration found for state `toto`/);

		expect(
			styles({
				// @ts-expect-error base state key is missing
				backgroundColor: {
					hover: "red",
				},
			}),
		).toBeTypeOf("string");

		expect(
			styles({
				// @ts-expect-error property value does not exist (native value)
				width: new Date(),
			}),
		).toBeTypeOf("string");

		expect(() =>
			styles({
				// @ts-expect-error property key does not exist
				nonExistingProperty: "unknownValue",
			}),
		).toThrow(/No configuration found for property `nonExistingProperty`/);

		expect(
			createVariants(styles, {
				color: {
					accent: { color: "warning" },
					// @ts-expect-error property value does not exist
					brand: { backgroundColor: "surfacePrimary" },
					neutral: { backgroundColor: "red", color: "lightgrey" },
				},
				size: {
					// @ts-expect-error property key does not exist
					medium: { height: 12 },
					small: { size: "auto", width: 100 },
				},
			}),
		).toBeTypeOf("function");

		expect(
			styles({
				color: "red",
				height: "100%",
				size: {
					base: "auto",
					hover: "100%",
				},
				width: "auto",
			}),
		).toBeTypeOf("string");
	});
});

const animationName = createKeyframes({
	25: {
		transform: "scale(1.25)",
	},
	"50%": {
		transform: "scale(1.5)",
	},
	"from": {
		transform: "scale(1)",
	},
	"to": {
		transform: "scale(1)",
	},
});

const styles = createStyles(
	{
		backgroundColor: true,
		color: true,
		padding: [6, 12, 18],
	},
	{
		states: {
			alt({ className, declaration }) {
				return `${className}[alt]{${declaration}}`;
			},
			hover({ className, declaration }) {
				return `${className}{${declaration}}`;
			},
		},
	},
);

const buttonVariants = createVariants(styles, {
	color: {
		accent: { backgroundColor: "lightcoral" },
		brand: { backgroundColor: "lightseagreen" },
		neutral: { backgroundColor: "lightgrey" },
	},
	size: {
		large: { padding: 18 },
		medium: { padding: 12 },
		small: { padding: 6 },
	},
});

const classNames = styles({
	backgroundColor: {
		alt: "lightgray",
		base: "lightcoral",
		hover: "lightcyan",
	},
	color: "lightblue",
	padding: {
		base: 6,
	},
});
