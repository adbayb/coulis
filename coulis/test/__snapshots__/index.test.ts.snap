// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`coulis > should extract styles 1`] = `
[
  {
    "attributes": {
      "data-coulis-cache": "c77e20e50,c8e4ce7ac",
      "data-coulis-id": "global",
    },
    "content": "@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}*,*::before,*::after{box-sizing:inherit;}.globalClass+.otherGlobalClass{border:1px solid black;border-radius:4px;}@charset "utf-8";@font-face{font-family:'AliasedHelvetica';src:local(Helvetica);}@import url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap');html{box-sizing:border-box;}html,body{background-color:lightcoral;font-family:Open Sans, AliasedHelvetica;}",
  },
  {
    "attributes": {
      "data-coulis-cache": "c2a100aa4,c6a0088d1,ccadc8a20",
      "data-coulis-id": "shorthand",
    },
    "content": ".c2a100aa4{padding:6px;}.c6a0088d1{padding:18px;}.ccadc8a20{padding:12px;}",
  },
  {
    "attributes": {
      "data-coulis-cache": "cdd299de8,cbb3938,ce31a3799,c18a19c73,c72f7daf8,c5ea183e8",
      "data-coulis-id": "longhand",
    },
    "content": ".cdd299de8[alt]{background-color:lightgray;}.cbb3938{background-color:lightcoral;}.ce31a3799{background-color:lightcyan;}.c18a19c73{color:lightblue;}.c72f7daf8{background-color:lightseagreen;}.c5ea183e8{background-color:lightgrey;}",
  },
  {
    "attributes": {
      "data-coulis-cache": "",
      "data-coulis-id": "atShorthand",
    },
    "content": "",
  },
  {
    "attributes": {
      "data-coulis-cache": "",
      "data-coulis-id": "atLonghand",
    },
    "content": "",
  },
]
`;

exports[`coulis > should extract styles given stringified styles 1`] = `"<style data-coulis-cache="c77e20e50,c8e4ce7ac" data-coulis-id="global">@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}*,*::before,*::after{box-sizing:inherit;}.globalClass+.otherGlobalClass{border:1px solid black;border-radius:4px;}@charset "utf-8";@font-face{font-family:'AliasedHelvetica';src:local(Helvetica);}@import url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap');html{box-sizing:border-box;}html,body{background-color:lightcoral;font-family:Open Sans, AliasedHelvetica;}</style><style data-coulis-cache="c2a100aa4,c6a0088d1,ccadc8a20" data-coulis-id="shorthand">.c2a100aa4{padding:6px;}.c6a0088d1{padding:18px;}.ccadc8a20{padding:12px;}</style><style data-coulis-cache="cdd299de8,cbb3938,ce31a3799,c18a19c73,c72f7daf8,c5ea183e8" data-coulis-id="longhand">.cdd299de8[alt]{background-color:lightgray;}.cbb3938{background-color:lightcoral;}.ce31a3799{background-color:lightcyan;}.c18a19c73{color:lightblue;}.c72f7daf8{background-color:lightseagreen;}.c5ea183e8{background-color:lightgrey;}</style><style data-coulis-cache="" data-coulis-id="atShorthand"></style><style data-coulis-cache="" data-coulis-id="atLonghand"></style>"`;
